### PR TITLE
ICDS Dashboard: Incorrect maps fill

### PR DESCRIPTION
**Pull request by saiprakash-v**
https://dimagi-dev.atlassian.net/browse/ICDS-1351
##### SUMMARY

data-maps library adds "Id" of the location from topojson as class to the corresponding canvas element. 
eg: South west Delhi has Id "South West". In this case, its added as two seperate classes ("South" and "West") to the canvas element. Also there is South Delhi, with id "South". Its canvas element has class "South".
Sample of map.data.data from api response looks like:
Screenshot from 2020-04-08 13-21-07

library traverse through each entry in the data and fill its corresponding colour as shown:
`for (var subunit in map.data.data) {`
`     svg.selectAll('.' + subunit).transition().style('fill', map.data.data.subunit.fillKey);`
` }`

So svg always select "South west Delhi" as well when "South Delhi was supposed to fill the color.

So I have created a unique class based on the ID and re-rendered map colors entirely.

### CODE REVIEW

File: custom/icds_reports/static/js/directives/indie-map/indie-map.directive.js [hunk 1]
```diff
@@ -191,6 +191,24 @@ function IndieMapController($scope, $compile, $location, $filter, storageService
...
+                $(function () {
+                    var svg = d3.select('#map svg');
+                    svg.selectAll(".datamaps-subunit").transition().style('fill', vm.map.fills.defaultFill);
+                    var locations = document.getElementsByClassName("datamaps-subunit");
+                    for (var i = 0; i < locations.length; i++) {
+                        var combinedClass = "";
+                        for (var j = 0; j < locations[i].classList.length; j++) {
+                            combinedClass += locations[i].classList[j];
+                        }
+                        locations[i].classList.add(combinedClass);
```
Comment (czue) [inline custom/icds_reports/static/js/directives/indie-map/indie-map.directive.js:203]: Ok so if I'm understanding this right, this basically takes `<div class="datamaps-subunit South Delhi">` and makes it `<div class="datamaps-subunit South Delhi datamaps-subunitSouthDelhi">`?

File: custom/icds_reports/static/js/directives/indie-map/indie-map.directive.js [hunk 2]
```diff
@@ -191,6 +191,24 @@ function IndieMapController($scope, $compile, $location, $filter, storageService
...
+                    }
+                    for (var locationId in vm.map.data) {
+                        if (vm.map.data.hasOwnProperty(locationId)) {
+                            svg.selectAll('.datamaps-subunit' + locationId.replace(/\s/g,''))
```
Comment (czue) [inline custom/icds_reports/static/js/directives/indie-map/indie-map.directive.js:207]: And then this manually sets the colors based on `locationId`, which is "South Delhi" and then you replace the whitespace with empty strings to generate a selector for the newly added `datamaps-subunitSouthDelhi` class?

File: custom/icds_reports/static/js/directives/indie-map/indie-map.directive.js [hunk 3]
```diff
@@ -191,6 +191,24 @@ function IndieMapController($scope, $compile, $location, $filter, storageService
                         .translate([element.offsetWidth / 2, element.offsetHeight / div]);
                     path = d3.geo.path().projection(projection);
                 }
+                $(function () {
+                    var svg = d3.select('#map svg');
+                    svg.selectAll(".datamaps-subunit").transition().style('fill', vm.map.fills.defaultFill);
+                    var locations = document.getElementsByClassName("datamaps-subunit");
+                    for (var i = 0; i < locations.length; i++) {
```
Comment (czue) [inline custom/icds_reports/static/js/directives/indie-map/indie-map.directive.js:198]: can you make this a function called e.g. `addCombinedSelectorClassToMaps`?

File: custom/icds_reports/static/js/directives/indie-map/indie-map.directive.js [hunk 4]
```diff
@@ -191,6 +191,24 @@ function IndieMapController($scope, $compile, $location, $filter, storageService
...
+                        var combinedClass = "";
+                        for (var j = 0; j < locations[i].classList.length; j++) {
+                            combinedClass += locations[i].classList[j];
+                        }
+                        locations[i].classList.add(combinedClass);
+                    }
+                    for (var locationId in vm.map.data) {
```
Comment (czue) [inline custom/icds_reports/static/js/directives/indie-map/indie-map.directive.js:205]: and make this a function called, e.g. `colorMapBasedOnCombinedSelectorClass`?

**Review by czue — COMMENTED** [review]
Really nice detective work figuring this out!

I had some minor code suggestions for this approach - but one bigger picture question - is it not possible to add metadata somewhere else to the "datamaps-subunit" so that we don't have to rely on this class/text manipulation? Guessing not, but just wanted to confirm that you looked into that option.

File: custom/icds_reports/static/js/directives/indie-map/indie-map.directive.js (hunk 2 above)
Comment (saiprakash-v) [reply custom/icds_reports/static/js/directives/indie-map/indie-map.directive.js:207]: Yes

**Comment by saiprakash-v** [comment]
> Really nice detective work figuring this out!
> 
> I had some minor code suggestions for this approach - but one bigger picture question - is it not possible to add metadata somewhere else to the "datamaps-subunit" so that we don't have to rely on this class/text manipulation? Guessing not, but just wanted to confirm that you looked into that option.

I did not get what you meant. Can you explain more?

**Comment by czue** [comment]
> I did not get what you meant. Can you explain more?

I was thinking like - if we were coding from scratch we'd maybe do something like `<div data-location-guid="{location.guid}">` and then in the function that updates the colors we could use that attribute instead of relying on the space-removal-from-the-name thing. But I'm not sure we have any flexibility of inserting additional attributes down in datamaps where those elements are being generated

**Review by czue — APPROVED** [review]

File: custom/icds_reports/static/js/directives/indie-map/indie-map.directive.js [hunk 5]
```diff
@@ -191,6 +191,12 @@ function IndieMapController($scope, $compile, $location, $filter, storageService
                         .translate([element.offsetWidth / 2, element.offsetHeight / div]);
                     path = d3.geo.path().projection(projection);
                 }
+                $(function () {
+                    var svg = d3.select('#map svg');
+                    svg.selectAll(".datamaps-subunit").transition().style('fill', vm.map.fills.defaultFill);
+                    vm.addCombinedSelectorClassToMaps(document.getElementsByClassName("datamaps-subunit"));
```
Comment (czue) [inline custom/icds_reports/static/js/directives/indie-map/indie-map.directive.js:197]: I was tempted to request a comment here explaining a bit more about this. feel free to add one in a follow up if you think it'd be helpful

Comment (saiprakash-v) [reply custom/icds_reports/static/js/directives/indie-map/indie-map.directive.js:197]: Sure

**Pull request merged**